src/CoreShop/Component/ProductQuantityPriceRules/Calculator/VolumeCalculator.php

```diff
@@ -0,0 +1,129 @@
+<?php
+/**
+ * CoreShop.
+ *
+ * This source file is subject to the GNU General Public License version 3 (GPLv3)
+ * For the full copyright and license information, please view the LICENSE.md and gpl-3.0.txt
+ * files that are distributed with this source code.
+ *
+ * @copyright  Copyright (c) 2015-2019 Dominik Pfaffenbauer (https://www.pfaffenbauer.at)
+ * @license    https://www.coreshop.org/license     GNU General Public License version 3 (GPLv3)
+ */
+
+namespace CoreShop\Component\ProductQuantityPriceRules\Calculator;
+
+use CoreShop\Component\ProductQuantityPriceRules\Model\ProductQuantityPriceRuleInterface;
+use CoreShop\Component\Registry\ServiceRegistryInterface;
+use CoreShop\Component\ProductQuantityPriceRules\Model\QuantityRangeInterface;
+use Doctrine\Common\Collections\Collection;
+use CoreShop\Component\ProductQuantityPriceRules\Model\QuantityRangePriceAwareInterface;
+use CoreShop\Component\ProductQuantityPriceRules\Rule\Action\ProductQuantityPriceRuleActionInterface;
+
+class VolumeCalculator implements CalculatorInterface
+{
+    /**
+     * @var ServiceRegistryInterface
+     */
+    protected $actionRegistry;
+
+    /**
+     * @param ServiceRegistryInterface $actionRegistry
+     */
+    public function __construct(ServiceRegistryInterface $actionRegistry)
+    {
+        $this->actionRegistry = $actionRegistry;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function calculatePerQuantity(
+        ProductQuantityPriceRuleInterface $quantityPriceRule,
+        QuantityRangePriceAwareInterface $subject,
+        int $quantity,
+        int $originalPrice,
+        array $context
+    ) {
+        $locatedRange = $this->locate($quantityPriceRule->getRanges(), $quantity);
+
+        if (!$locatedRange instanceof QuantityRangeInterface) {
+            return false;
+        }
+
+        $price = $this->calculateRangePrice($locatedRange, $subject, $originalPrice, $context);
+
+        return !is_numeric($price) || $price === 0 ? false : $price;
+
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function calculatePerItem(
```

Comment:
Shouldn't we split that into two interfaces? How should I know what calculator is capable of when it just throws an exception?